### PR TITLE
Extract data on the fly in packaged builders

### PR DESCRIPTION
**Pull request by mariosasko**
Instead of waiting for data files to be extracted in the packaged builders, we can prepend the compression prefix and extract them as they are being read (using `fsspec`). This saves disk space (deleting extracted archives is not set by default) and slightly speeds up dataset generation (less disk reads)

### CODE REVIEW

**Comment by mariosasko** [comment]
CI failures are unrelated, so this is ready for the review

**Review by lhoestq — COMMENTED** [review]
Nice ! just one comment:

File: src/datasets/download/download_manager.py [hunk 1]
```diff
@@ -503,7 +324,28 @@ def iter_files(self, paths: Union[str, List[str]]):
         >>> files = dl_manager.iter_files(files)
         ```
         """
-        return FilesIterable.from_paths(paths)
+        return FilesIterable.from_urlpaths(paths)
+
+    def _extract_on_the_fly(self, path: str, download_config: DownloadConfig) -> str:
```
Comment (lhoestq) [inline src/datasets/download/download_manager.py:329]: Maybe this logic can be in `cached_path` instead ?

File: tests/fixtures/fsspec.py [hunk 2]
```diff
@@ -107,6 +107,8 @@ def mockfs(tmp_path_factory, mock_fsspec):
 
 @pytest.fixture
 def tmpfs(tmp_path_factory, mock_fsspec):
+    print("Executed" * 10)
+    print(_fsspec_registry["tmp"])
```
Comment (lhoestq) [inline tests/fixtures/fsspec.py:111]:
```suggestion
```

File: src/datasets/utils/file_utils.py [hunk 3]
```diff
@@ -217,10 +228,30 @@ def cached_path(
...
     if download_config.extract_compressed_file:
+        if download_config.extract_on_the_fly:
+            # Add a compression prefix to the compressed file so that it can be extracted
+            # as it's being read using xopen.
+            protocol = _get_extraction_protocol(output_path, download_config=download_config)
+            extension = _get_path_extension(url_or_filename)
+            if (
+                protocol
+                and extension not in ["tgz", "tar"]
+                and not url_or_filename.endswith((".tar.gz", ".tar.bz2", ".tar.xz"))
```
Comment (lhoestq) [inline src/datasets/utils/file_utils.py:239]:
```suggestion
                and not url_or_filename.split("::")[0].endswith((".tar.gz", ".tar.bz2", ".tar.xz"))
```

File: src/datasets/utils/file_utils.py [hunk 4]
```diff
@@ -217,10 +228,30 @@ def cached_path(
         return output_path
 
     if download_config.extract_compressed_file:
+        if download_config.extract_on_the_fly:
+            # Add a compression prefix to the compressed file so that it can be extracted
+            # as it's being read using xopen.
+            protocol = _get_extraction_protocol(output_path, download_config=download_config)
+            extension = _get_path_extension(url_or_filename)
```
Comment (lhoestq) [inline src/datasets/utils/file_utils.py:235]:
```suggestion
            extension = _get_path_extension(url_or_filename.split("::")[0])
```

**Review by lhoestq — APPROVED** [review]
LGTM !

**Pull request merged**